core/src/ast/macros.rs

```diff
@@ -0,0 +1,302 @@
+use std::{collections::BTreeMap, fmt::Debug};
+
+use proc_macro2::{TokenStream, TokenTree};
+use quote::{ToTokens, TokenStreamExt};
+use syn::{
+    braced, bracketed,
+    buffer::{Cursor, TokenBuffer},
+    parenthesized,
+    parse::{self, Parse},
+    token, Error, Expr, Ident, ImplItem, ImplItemMacro, Item, ItemMacro, Token,
+};
+
+#[derive(Default)]
+pub struct Macros {
+    defs: BTreeMap<Ident, MacroRules>,
+}
+
+impl Macros {
+    pub fn new() -> Macros {
+        Macros {
+            defs: BTreeMap::new(),
+        }
+    }
+
+    pub fn read_item_macro(&mut self, input: &ItemMacro) -> Option<Vec<Item>> {
+        let mac = Macro::from_syn(input);
+        if let Ok((ident, mac)) = mac {
+            match mac {
+                Macro::MacroRules(rules) => {
+                    self.defs.insert(ident, rules);
+                    // Macro rules add no new items:
+                    None
+                }
+                Macro::MacroMatch(matched) => {
```

Comment:
issue: I'd prefer a cleaner `add_macro` / `evaluate_macro` separation, with asserts about the right type of `Macro` coming in